design/rest-api-guideline.md

```diff
@@ -0,0 +1,389 @@
+
+
+  - [Principles](#principles)
+    - [Level of Service Maturity](#level-of-service-maturity)
+    - [Resource Path](#resource-path)
+      - [Group](#group)
+      - [Versioning](#versioning)
+      - [Extensions](#extensions)
+      - [Technology Preview](#technology-preview)
+      - [Store Resources](#store-resources)
+      - [Controller Resources](#controller-resources)
+    - [Resource Representation](#resource-representation)
+    - [HTTP Status Codes](#http-status-codes)
+    - [HTTP Verbs](#http-verbs)
+      - [GET](#get)
+      - [POST](#post)
+      - [PUT](#put)
+      - [DELETE](#delete)
+  - [Practices](#practices)
+    - [Error Handling](#error-handling)
+    - [Filtering](#filtering)
+    - [Pagination](#pagination)
+    - [Documentation](#documentation)
+  - [Examples](#examples)
+
+## Principles
+
+This section defines the general principles that should be considered by API developers, regardless of the practices defined through this document.
+
+### Level of Service Maturity
+
+Based on [Richardson Maturity Model](https://martinfowler.com/articles/richardsonMaturityModel.html), developers MUST adhere to the maturity Level 2, where interactions with the API endpoints are based on the usage of HTTP Verbs and status codes to communicate status and errors.
+
+### Resource Path
+
+In Keycloak, we choose to stick with the following convention for resource paths:
+
+```
+/apis/{API_GROUP}/{version}
+```
+
+When the resource operates in the context of a `realm`, the resource path should be as follows:
+
+```
+/apis/{API_GROUP}/{version}/realms/{realm}
+```
+
+When operating in the context of a `realm`, a `/realms/{realm}` should be included at the end of the resource root path.
+
+#### Group
+
+The `API_GROUP` refers to a set of one or more functionality that can be grouped together. The endpoints associated with a group can manage different resources as far as they are strictly related with the functionality represented by the group.
+
+By including the group in the resource path, we aim to provide a clear and consistency view of the functionalities exposed by the different APIs, with more control over the groups of functionality (in terms of functional and non-functional aspects) as well as make more clear how we communicate the functional structure of these APIs through the documentation.
+
+Here is an example of a resource path related with the group of functionalities provided by the Account Management REST API:
+
+```
+/apis/accounts/v1/{user_id}
+```
+
+The path refers to the `accounts` group and any resource/operation related with user account management (in this case for user with id `user_id`) can be accessed from that path.
+Note that by grouping functionalities together and in conjunction with the versioning schema, we can keep the root path and also the semantics of the API without impact the adaptability for new changes or improvements to the API.
+
+Group names should be in lowercase, a single word name and plural.
```

Comment:
As described, the group should represent a group of functionalities. These functionalities will probably operate on a top-level resource, even this top-level resource is logical.

The admin API example you mentioned is exactly what we are trying to improve. With the resource path format we are proposing, any Admin API we expose (e.g: users, groups, clients, identity providers, realms, etc) will have distinct groups and not include "admin" in any part of the resource path.

The path format herein proposed aims to create a uniform view of our APIs while keeping consistency on how they are exposed. Note that "/apis/admins" does not really provide any meaning for consumers, what they are interested in is how to manage resources. For instance, for users "/apis/users".

if you see the example section, you'll see how the User API should be exposed.

Reply:
Ok, I think I get the idea now - the API "group" is basically a resource (in terms of REST). But then why to call it a group when it's a resource? Isn't it a bit confusing? :)

One thing I don't understand now is why the API version is present after the resource (`/apis/{API_GROUP}/{version}`)? Are we going to have a different versioning for each resource? Isn't it a best practice to have a single unified version for the whole API (to not overcomplicate things)?

Reply:
Yeah, the group is a root resource which will probably have correlate subresources. So, a group of resources. If you have a better word for this, I'm happy to change it.

Regarding versioning, let me give you a concrete example. As you know we have a v2 of the account console which will become the official version very soon. Previously we had v1 which was not backed by a REST API, but if it were, the versioning scheme on a per-group/root resource basis would help. I think this model is quite flexible since we have different APIs that may evolve each one on its own pace. See, most of the time the version will be the same across all APIs, at least until some point in time, but when this is not true anymore we have the possibility to tell clients about breaking changes and new functionality being added for a specific API.

Reply:
I get your point but let me give you another example. :) Keycloak components. We have a unified version for all components in the server and (currently) even adapters as well. We don't have a different version for each SPI.
And I'm not sure it's very RESTful to put a version after (root) resource. (Personally, I haven't seen anyone else taking this approach. :))

Reply:
Good example :) That is true and you will probably agree on about how much we are careful to change these SPIs.

The model I'm using is pretty much similar to how kubernetes or even google. For the latter, you may argue that each API is indeed a separated thing but that is the idea behind these groups.

If you want more details about how k8s look this https://github.com/kubernetes/community/blob/master/contributors/design-proposals/api-machinery/api-group.md.

Reply:
I believe there will be multiple APIs (API group). I don't see these as a top-level resource, but rather a specific API. Admin is one, account is another, etc.. So I would argue it should not be plural (admin, account).

Reply:
@stianst +1
It then imho makes sense to have independent version for each API.

Reply:
Based on the idea of the guideline, API groups can be seen as a top-level resource. Thus the reason for having a plural name.

There are some examples in guideline for the account case.

Regardgim Admin, that is something I'm trying to avoid, and I'm happy to make this more clear. All sub-resources that we have today for things like users, etc, will have their own api group instead of subresources of an API group such as 'realms' or 'admin'. Instead of `/realms/{realm}/users` we would have just `/apis/users`.

The "admin" is a logical group and not really part of the API path.